File: src/dynamic-programming/dungeon-game.ts
export function calculateMinimumHP(dungeon: number[][]): number {
  const n = dungeon.length
  const m = dungeon[0].length
  //dp[i][j]代表从坐标(i,j)走到终点需要的最小生命值
  const dp: number[][] = Array.from({ length: n + 1 }, () =>
    Array.from<number>({ length: m + 1 }).fill(Infinity)
  )

  dp[n - 1][m] = dp[n][m - 1] = 1

  for (let i = n - 1; i >= 0; --i) {
    for (let j = m - 1; j >= 0; --j) {
      const minin = Math.min(dp[i + 1][j], dp[i][j + 1])

      dp[i][j] = Math.max(minin - dungeon[i][j], 1)
    }
  }

  return dp[0][0]
}

File: src/dynamic-programming/dungeon-game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int n = dungeon.size();
        int m = dungeon[0].size();
        //dp[i][j]表示从[i,j]走到右下角需要的最小初始HP
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, INT_MAX));
        dp[n-1][m] = dp[n][m-1] = 1;

        for(int i = n - 1; i >= 0; --i) {
            for(int j = m - 1; j >= 0; --j) {
                int minin = min(dp[i + 1][j], dp[i][j + 1]);

                dp[i][j] = max(minin - dungeon[i][j], 1);
            }
        }

        return dp[0][0];
    }
};